stablehlo/dialect/StablehloOps.cpp

```diff
@@ -5612,66 +5612,49 @@ LogicalResult ScatterOp::verify() {
-             << condReturnOp->getOperand(0).getType();
-  }
-  // Check the return type for the body block.
-  {
-    auto bodyReturnOp = cast<ReturnOp>(getBody().front().back());
-    if (bodyReturnOp->getNumOperands() != getNumOperands())
-      return bodyReturnOp.emitOpError()
-             << "expects body to return a many value as the operands ("
-             << getNumOperands() << "), got " << bodyReturnOp->getNumOperands();
-    for (const auto& enumeratedOperandTypes : llvm::enumerate(
-             llvm::zip(bodyReturnOp->getOperandTypes(), getOperandTypes()))) {
-      Type operandType = std::get<0>(enumeratedOperandTypes.value());
-      Type returnType = std::get<1>(enumeratedOperandTypes.value());
-      if (operandType != returnType)
-        return bodyReturnOp.emitOpError()
-               << "type mismatch between operand #"
-               << enumeratedOperandTypes.index()
-               << " and the enclosing WhileOp returned value: " << operandType
-               << " vs " << returnType;
-    }
-  }
+LogicalResult WhileOp::inferReturnTypeComponents(
+    MLIRContext*, Optional<Location> location, ValueShapeRange operands,
+    DictionaryAttr attributes, RegionRange regions,
+    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes) {
+  WhileOp::Adaptor adaptor(operands, attributes, regions);
+  auto operandTypes = adaptor.getOperand().getTypes();
+  auto condArgsTypes = adaptor.getCond().front().getArgumentTypes();
+  auto bodyArgsTypes = adaptor.getBody().front().getArgumentTypes();
+  auto bodyReturnTypes =
+      cast<ReturnOp>(adaptor.getBody().front().getTerminator())
+          ->getOperandTypes();
+  if (!hlo::isCompatibleForHloTypeInference(operandTypes, condArgsTypes))
+    return emitOptionalError(location,
+                             "expect operands are compatible with condition "
+                             "block arguments but got ",
+                             operandTypes, " vs ", condArgsTypes);
+  if (!hlo::isCompatibleForHloTypeInference(operandTypes, bodyArgsTypes))
+    return emitOptionalError(
+        location,
+        "expect operands are compatible with body block arguments but got ",
+        operandTypes, " vs ", bodyArgsTypes);
+  if (!hlo::isCompatibleForHloTypeInference(operandTypes, bodyReturnTypes))
+    return emitOptionalError(
+        location,
+        "expect operands are compatible with body block return types but got ",
+        operandTypes, " vs ", bodyReturnTypes);
+
+  auto condReturnTypes =
+      cast<ReturnOp>(adaptor.getCond().front().back()).getOperandTypes();
+  if (condReturnTypes.size() != 1)
+    return emitOptionalError(
+        location, "expect condition body returns a single value but got ",
+        condReturnTypes.size());
+  auto operandType = condReturnTypes[0].dyn_cast<RankedTensorType>();
+  if (operandType && operandType.getRank() != 0 ||
+      !getElementTypeOrSelf(condReturnTypes[0]).isInteger(1))
```

Comment:
I think that `getElementTypeOrSelf` looks awkward because it seems to suggest that `condReturnTypes[0]` might not be a tensor type which cannot happen. How about rewriting this as follows:

```
  auto operandType = condReturnTypes[0].dyn_cast<TensorType>();
  if ((operandType.hasRank() && operandType.getRank() != 0) ||
      !operandType.getElementType().isInteger(1))
    ...
```

Reply:
Looks nice! Fixed.